src/spider/client/Data.hpp

```diff
@@ -0,0 +1,81 @@
+#ifndef SPIDER_CLIENT_DATA_HPP
+#define SPIDER_CLIENT_DATA_HPP
+
+#include <functional>
+#include <memory>
+#include <string>
+#include <vector>
+
+namespace spider {
+
+class DataImpl;
+
+template <class T>
+class Data {
+private:
+    std::unique_ptr<DataImpl> m_impl;
+
+public:
+    /**
+     * Gets the values stored in Data.
+     * @return value stored in Data.
+     */
+    auto get() -> T;
+    /**
+     * Indicates that the data is persisted and should not be rollbacked
+     * on failure recovery.
+     */
+    // Not implemented in milestone 1
+    // void mark_persist();
+    /**
+     * Sets locality list of the data.
+     * @param nodes nodes that has locality
+     * @param hard true if the locality list is a hard requirement, false otherwise
+     */
+    void set_locality(std::vector<std::string> const& nodes, bool hard);
```

Comment:
* We prefer to use std::span and std::string_view instead of const vectors or strings.
* Why do we need this method compared to the one in the builder?

Reply:
Locality information might not be available at the time we create the data. For example, we create an associated data for HDFS files before we create the actual HDFS files and write to them so that Spider can gc the HDFS files on failure during write. However, we don't know the locality of the HDFS files before creating them, and thus we need a way to set locality on the fly.